### PR TITLE
Add tagging to messages

### PR DESCRIPTION
**Pull request by miaulalala**
As discussed in #25

### CODE REVIEW

File: lib/Db/Tag.php [hunk 1]
```diff
@@ -0,0 +1,61 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * @copyright 2019 Christoph Wurst <christoph@winzerhof-wurst.at>
```
Comment (ChristophWurst) [inline lib/Db/Tag.php:6]: It wasn't me

File: lib/Db/TagMapper.php [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+					break;
+				case $i === 5:
+					$tag->setDisplayName('Later');
+					$tag->setColor('#800080');
+					$tag->setIsSystemFlag(1);
+					break;
+			}
+			try {
+				$this->insert($tag);
+			} catch (\Doctrine\DBAL\Exception\UniqueConstraintViolationException $e) {
```
Comment (ChristophWurst) [inline lib/Db/TagMapper.php:182]: I don't think you can actually catch the Doctrine exception. Nextcloud will wrap that exception with its own. We can catch the \OCP\Db\Exception instead and check the reason

File: lib/Db/TagMapper.php [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+				$this->logger->debug($e->getMessage());
```
Comment (ChristophWurst) [inline lib/Db/TagMapper.php:183]: when logging an exception pass the object to the logger, so the trace is included in the log line.

```suggestion
				$this->logger->debug($e->getMessage(), ['exception' => $e]);
```

File: lib/Migration/Version1100Date20210304143008.php [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+use Closure;
+use OCA\Mail\Db\MailAccountMapper;
+use OCA\Mail\Db\TagMapper;
+use OCP\DB\ISchemaWrapper;
+use OCP\Migration\IOutput;
+use OCP\Migration\SimpleMigrationStep;
+
+/**
+ * Auto-generated migration step.
```
Comment (ChristophWurst) [inline lib/Migration/Version1100Date20210304143008.php:15]: you can drop this note

File: src/store/actions.js [hunk 5]
```diff
@@ -556,9 +558,6 @@ export default {
 		} finally {
 			showNewMessagesNotification(newMessages)
 		}
-	},
-	toggleEnvelopeFlagged({ commit, getters }, envelope) {
```
Comment (ChristophWurst) [inline src/store/actions.js:560]: I think those three lines should be restored

File: lib/Db/Tag.php (hunk 1 above)
Comment (miaulalala) [reply lib/Db/Tag.php:6]: That's copypasta for you ;)

File: lib/Db/TagMapper.php (hunk 2 above)
Comment (miaulalala) [reply lib/Db/TagMapper.php:182]: I tried the exception but the UniqueConstraintViolationException doesn't seem to get wrapped.

File: src/store/actions.js (hunk 5 above)
Comment (miaulalala) [reply src/store/actions.js:560]: 😬

File: lib/Db/TagMapper.php [hunk 6]
```diff
@@ -130,10 +142,37 @@ public function untagMessage(Tag $tag, string $messageId) {
 		$qb->execute();
 	}
 
+	/**
+	 * @param array $messages
+	 * @return array [[<jshdkfjhdsjhfdskjhfdkjshjkdshfjkdsh@test.com>] => [1, 2]]
```
Comment (miaulalala) [inline lib/Db/TagMapper.php:147]: no longer accurate

**Comment by ChristophWurst** [comment]
- [ ] Bump the version in appinfo/info.xml from alpha.1 to alpha.2. This will then trigger Nextcloud to check for new migration files and devs pulling master after we merge this PR will be migrated.

File: lib/Service/SetupService.php [hunk 7]
```diff
@@ -139,6 +150,8 @@ public function createNewAccount($accountName, $emailAddress, $imapHost, $imapPo
 		$this->accountService->save($newAccount);
 		$this->logger->debug("account created " . $newAccount->getId());
 
+		$this->tagMapper->createDefaultTags($newAccount->getMailAccount());
```
Comment (ChristophWurst) [inline lib/Service/SetupService.php:153]:
```suggestion
		$this->tagMapper->createDefaultTags($newAccount);
```

File: lib/Db/TagMapper.php [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+	 * @link https://github.com/nextcloud/mail/issues/25
+	 */
+	public function createDefaultTags(MailAccount $account) {
+		for ($i = 1; $i < 6; $i++) {
+			$tag = new \OCA\Mail\Db\Tag();
+			$tag->setImapLabel('$label' . $i);
+			$tag->setUserId($account->getUserId());
+			switch ($i) {
+				case $i === 1:
+					$tag->setDisplayName('Important');
```
Comment (ChristophWurst) [inline lib/Db/TagMapper.php:190]: We should localize the labels here

**Review by ChristophWurst — DISMISSED** [review]
Some little quirks I found. Otherwise reading the tags from Thunderbird into our DB works as expected :)

I noticed this as well:

Bildschirmfoto von 2021-03-16 09-51-09

(accidentally added the same account multiple times)

Is that a problem? Checking before insert is likely expensive, so we might accept the cost of duplication.

**Comment by ChristophWurst** [comment]
I'm not a fan of the naming *system tags* because they have an existing meaning in Nextcloud for file tagging. How about we rename to *default tags*?

**Comment by ChristophWurst** [comment]
- [ ] BUG when marking messages as important it should add the tag `$label1` now

instead I currently see this being inserted 
Bildschirmfoto von 2021-03-16 10-14-49

**Comment by miaulalala** [comment]
> I'm not a fan of the naming _system tags_ because they have an existing meaning in Nextcloud for file tagging. How about we rename to _default tags_?

I agree, done


> Some little quirks I found. Otherwise reading the tags from Thunderbird into our DB works as expected :)
> 
> I noticed this as well:
> 
> Bildschirmfoto von 2021-03-16 09-51-09
> 
> (accidentally added the same account multiple times)
> 
> Is that a problem? Checking before insert is likely expensive, so we might accept the cost of duplication.

That's one issue with cleanup - not deleting the records when removing the account will duplicate the entries on the initial sync when readding. Duplicates are fine though. When removing the tags it should clean up all orphaned records with the same tag, as well. If this is an account for an existing user, the tags aren't newly generated in `oc_mail_tags` so they should have the same database id for the user.

File: lib/Db/TagMapper.php [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+	 * @return Entity
+	 *
+	 * @throws DoesNotExistException
+	 */
+	public function getTagByImapLabel(string $imapLabel, string $userId): Entity {
+		$qb = $this->db->getQueryBuilder();
+		$qb->select('*')
+			->from($this->getTableName())
+			->where(
+				$qb->expr()->like('imap_label', $qb->createNamedParameter($imapLabel)),
```
Comment (ChristophWurst) [inline lib/Db/TagMapper.php:67]:
```suggestion
				$qb->expr()->eq('imap_label', $qb->createNamedParameter($imapLabel)),
```

right?

File: lib/Model/IMAPMessage.php [hunk 10]
```diff
@@ -673,7 +674,13 @@ public function setInReplyTo(string $message) {
...
 	}
 
-	public function toDbMessage(int $mailboxId): \OCA\Mail\Db\Message {
+	/**
+	 * Cast all values from an IMAP message into the correct DB format
+	 *
+	 * @param integer $mailboxId
+	 * @return \OCA\Mail\Db\Message
+	 */
+	public function toDbMessage(int $mailboxId, MailAccount $account): \OCA\Mail\Db\Message {
```
Comment (ChristophWurst) [inline lib/Model/IMAPMessage.php:683]: import the `Message` class

File: lib/Model/IMAPMessage.php [hunk 11]
```diff
@@ -707,10 +714,57 @@ public function toDbMessage(int $mailboxId): \OCA\Mail\Db\Message {
...
+		$allowed = [
+			Horde_Imap_Client::FLAG_SEEN,
+			Horde_Imap_Client::FLAG_ANSWERED,
+			Horde_Imap_Client::FLAG_FLAGGED,
+			Horde_Imap_Client::FLAG_DELETED,
+			Horde_Imap_Client::FLAG_DRAFT,
+			Horde_Imap_Client::FLAG_RECENT,
+			Horde_Imap_Client::FLAG_JUNK,
+			'junk',
+			'$junk',
```
Comment (ChristophWurst) [inline lib/Model/IMAPMessage.php:731]: this is identical to `Horde_Imap_Client::FLAG_JUNK`

File: lib/Model/IMAPMessage.php [hunk 12]
```diff
@@ -707,10 +714,57 @@ public function toDbMessage(int $mailboxId): \OCA\Mail\Db\Message {
...
+			'$seen',
```
Comment (ChristophWurst) [inline lib/Model/IMAPMessage.php:732]: is this really an IMAP standard flag?

File: lib/Model/IMAPMessage.php [hunk 13]
```diff
@@ -707,10 +714,57 @@ public function toDbMessage(int $mailboxId): \OCA\Mail\Db\Message {
...
+	 *
+	 * @link https://github.com/nextcloud/mail/issues/25
+	 *
+	 * @param array $tags
+	 * @return array
+	 */
+	private function generateTagEntites(array $tags, string $userId): array {
+		$t = [];
+		foreach ($tags as $keyword) {
+			$tag = new \OCA\Mail\Db\Tag();
```
Comment (ChristophWurst) [inline lib/Model/IMAPMessage.php:762]: import

File: lib/Service/MailManager.php [hunk 14]
```diff
@@ -419,6 +427,50 @@ public function flagMessage(Account $account, string $mailbox, int $uid, string
...
+				}
+			} catch (Horde_Imap_Client_Exception $e) {
+				throw new ServiceException(
+					"Could not set message keyword on IMAP: " . $e->getMessage(),
+					(int) $e->getCode(),
+					$e
+				);
+			}
+		}
+		if ($value === true) {
```
Comment (ChristophWurst) [inline lib/Service/MailManager.php:467]:
```suggestion
		if ($value) {
```

it's a bool as per typehint

**Review by ChristophWurst — CHANGES_REQUESTED** [review]

File: lib/Model/IMAPMessage.php (hunk 12 above)
Comment (miaulalala) [reply lib/Model/IMAPMessage.php:732]: no but it's gmail's $seen flag - do we want to make a tag out of it?

File: lib/Db/TagMapper.php (hunk 2 above)
Comment (ChristophWurst) [reply lib/Db/TagMapper.php:227]: https://github.com/nextcloud/server/pull/26182

**Comment by ChristophWurst** [comment]
CI is partly failing because of https://github.com/codecov/codecov-action/issues/234

File: lib/Controller/MessagesController.php [hunk 15]
```diff
@@ -132,7 +131,8 @@ public function __construct(string $appName,
 								IURLGenerator $urlGenerator,
 								ContentSecurityPolicyNonceManager $nonceManager,
 								ITrustedSenderService $trustedSenderService,
-								IMailTransmission $mailTransmission) {
+								IMailTransmission $mailTransmission,
```
Comment (ChristophWurst) [inline lib/Controller/MessagesController.php:134]: let's revert those two lines. the trailing comma is problematic on php7.2 (an old bug installed new Mail in the past, hence we try to keep php7.2 for a little longer)

File: lib/Db/Message.php [hunk 16]
```diff
@@ -199,6 +203,20 @@ public function setTo(AddressList $to): void {
...
 
+	/**
+	 * @return array
+	 */
+	public function getTags(): array {
+		return $this->tags;
+	}
+
+	/**
+	 * @param array $tags
```
Comment (ChristophWurst) [inline lib/Db/Message.php:214]: more specific type please

File: lib/Db/Message.php [hunk 17]
```diff
@@ -199,6 +203,20 @@ public function setTo(AddressList $to): void {
 		$this->to = $to;
 	}
 
+	/**
+	 * @return array
```
Comment (ChristophWurst) [inline lib/Db/Message.php:207]: more specific type please

File: lib/Db/Message.php [hunk 18]
```diff
@@ -232,11 +250,14 @@ public function setFlag(string $flag, bool $value = true) {
 			// Ignore
 			return;
 		}
-
-		$this->setter(
-			$this->columnToProperty("flag_$flag"),
-			[$value]
-		);
+		if ($flag === '$label1') {
```
Comment (ChristophWurst) [inline lib/Db/Message.php:253]: how about a `\OCA\Mail\Db\Tag::LABEL_IMPORTANT` for better readability and lower chances of typos?

File: lib/Db/MessageMapper.php [hunk 19]
```diff
@@ -309,19 +326,26 @@ public function insertBulk(Message ...$messages): void {
...
 		$this->db->commit();
 	}
 
 	/**
-	 * @param Message ...$messages
-	 *
-	 * @return Message[]
+	 * @param Account $account
+	 * @param Message[] $messages
+	 * @return array
```
Comment (ChristophWurst) [inline lib/Db/MessageMapper.php:340]:
```suggestion
	 * @return Message[]
```

as it was before

File: lib/Db/MessageMapper.php [hunk 20]
```diff
@@ -360,13 +386,59 @@ public function updateBulk(Message ...$messages): array {
...
+				continue;
+			}
+
+			$toRemove = array_udiff($dbTags, $imapTags, function (Tag $a, Tag $b) {
+				return strcmp($a->getImapLabel(), $b->getImapLabel());
+			});
+			foreach ($toRemove as $tag) {
+				//add logging
+				$this->tagMapper->untagMessage($tag, $message->getMessageId());
+				$perf->step("Untagged message");
```
Comment (ChristophWurst) [inline lib/Db/MessageMapper.php:420]: I'd move that outside the loop so it only gets logged once

File: lib/Db/TagMapper.php [hunk 21]
```diff
@@ -0,0 +1,234 @@
...
+	 * @return Entity
+	 *
+	 * @throws DoesNotExistException
+	 */
+	public function getTag(int $id): Entity {
+		$qb = $this->db->getQueryBuilder();
+		$qb->select('*')
+			->from($this->getTableName())
+			->where($qb->expr()->eq('id', $qb->createNamedParameter($id, IQueryBuilder::PARAM_INT)));
+		$result = $this->findEntity($qb);
```
Comment (ChristophWurst) [inline lib/Db/TagMapper.php:107]: inline `$result`

File: lib/Db/TagMapper.php [hunk 22]
```diff
@@ -0,0 +1,234 @@
...
+				$qb->expr()->in('imap_message_id', $qb->createNamedParameter($ids, IQueryBuilder::PARAM_STR_ARRAY))
+			);
+		$qb = $qb->execute();
+		$queryResult = $qb->fetchAll();
+		if (empty($queryResult)) {
+			return [];
+		}
+
+		foreach ($queryResult as $qr) {
+			$result[$qr['imap_message_id']][] = $this->getTag((int)$qr['tag_id']);
```
Comment (ChristophWurst) [inline lib/Db/TagMapper.php:174]: `$result` is uninitalized and so is the index `$qr['imap_message_id']` when you first access it

File: lib/Service/MailManager.php [hunk 23]
```diff
@@ -23,34 +23,36 @@
...
 use OCA\Mail\Db\Mailbox;
-use OCA\Mail\Db\MailboxMapper;
 use OCA\Mail\Db\Message;
-use OCA\Mail\Db\MessageMapper as DbMessageMapper;
-use OCA\Mail\Events\BeforeMessageDeletedEvent;
+use function array_values;
+use OCA\Mail\Db\TagMapper;
+use Psr\Log\LoggerInterface;
+use Horde_Imap_Client_Socket;
+use OCA\Mail\Db\MailboxMapper;
```
Comment (ChristophWurst) [inline lib/Service/MailManager.php:37]: can we keep the imports sorted alphabetically?

File: lib/Service/Search/Flag.php [hunk 24]
```diff
@@ -34,7 +34,7 @@ class Flag {
 	public const ANSWERED = Horde_Imap_Client::FLAG_ANSWERED;
 	public const SEEN = Horde_Imap_Client::FLAG_SEEN;
 	public const FLAGGED = Horde_Imap_Client::FLAG_FLAGGED;
-	public const IMPORTANT = '\\important';
+	public const IMPORTANT = '$label1';
```
Comment (ChristophWurst) [inline lib/Service/Search/Flag.php:37]: mark as deprecated now that we know *important* is rather a *tag* than a *flag*

File: lib/Service/SetupService.php [hunk 25]
```diff
@@ -56,28 +57,36 @@ class SetupService {
...
+
 	public function __construct(AutoConfig $autoConfig,
 								AccountService $accountService,
 								ICrypto $crypto,
 								SmtpClientFactory $smtpClientFactory,
 								IMAPClientFactory $imapClientFactory,
-								LoggerInterface $logger) {
+								LoggerInterface $logger,
+								TagMapper $tagMapper
+								) {
```
Comment (ChristophWurst) [inline lib/Service/SetupService.php:73]: closing bracket should move  to previous line

**Review by ChristophWurst — CHANGES_REQUESTED** [review]

File: lib/Service/MailManager.php (hunk 23 above)
Comment (miaulalala) [reply lib/Service/MailManager.php:37]: I'm not sure why my IDE is doing that, I'm sorry! I'll see which extension is causing this.

**Comment by ChristophWurst** [comment]
- [ ] BUG message tags are always on the front-end

in `\OCA\Mail\Db\Message::jsonSerialize` you use `\OCA\Mail\Db\Message::getTags` but the tags prop is only set when when read messages from imap.

We need something similar to `\OCA\Mail\Db\MessageMapper::findRecipients` in the message mapper that loads each message's tags.

**Comment by ChristophWurst** [comment]
> miaulalala force-pushed the enhancement/tagging-messages branch

unless there is a change on master that we need here I would recommend to only do fixups and wait with the rebase until we have positive review & CI result. That will make it easier for me and others to check out the branch and pull the latest changes.

**Comment by ChristophWurst** [comment]
- [ ] BUG our old `$important` tag in inserted as a new tag, but we should just map it to `$label1`

Bildschirmfoto von 2021-03-18 10-30-17

**Comment by ChristophWurst** [comment]
- [ ] BUG I could have happened because of the way I reset my db before I re-migrated to your new schema changes here but this shouldn't happen:

Bildschirmfoto von 2021-03-22 12-10-50

**Comment by ChristophWurst** [comment]
- [ ] BUG `message.tags` is not an indexed object, but used as one

Bildschirmfoto von 2021-03-22 12-13-07

we have code that does `envelope.tags.$label1` but this only works it `envelope.tags` is an object with a key `$label1`. when we load messages/envelopes they have an *array* of tags, however. So we have to use something like a `envelope.tags.some(t => t.keyword === '$label1')` with a O(n) or we store the tags as a simpler object of keywords, like `{ '$label1': { id: 123, keyword: '$label1', displayName: 'Important' }` and then we can (again) use `envelope.tags.$label1` with O(1).

To sum up we should index tags by their keyword.

File: lib/Db/Tag.php [hunk 26]
```diff
@@ -0,0 +1,68 @@
...
+	 * @return array
+	 */
+	public function jsonSerialize() {
+		return [
+			'databaseId' => $this->getId(),
+			'userId' => $this->getUserId(),
+			'displayName' => $this->getDisplayName(),
+			'imapLabel' => $this->getImapLabel(),
+			'color' => $this->getColor(),
+			'isDefaultTag' => $this->getisDefaultTag(),
```
Comment (ChristophWurst) [inline lib/Db/Tag.php:65]:
```suggestion
			'isDefaultTag' => $this->getIsDefaultTag(),
```

**Review by ChristophWurst — CHANGES_REQUESTED** [review]

File: lib/Db/TagMapper.php [hunk 27]
```diff
@@ -0,0 +1,250 @@
...
+	 * @param array $messages
+	 * @return array
+	 */
+	public function getAllTagsForMessages(array $messages):array {
+		$ids = array_map(function (Message $message) {
+			return $message->getMessageId();
+		}, $messages);
+
+		$qb = $this->db->getQueryBuilder();
+		$ids = $qb->select('mt.*')
```
Comment (ChristophWurst) [inline lib/Db/TagMapper.php:175]: `$ids` is never used

Comment (ChristophWurst) [reply lib/Db/TagMapper.php:175]: did you mean to call `$ids->execute()` below? let's rename to `selectIdsQuery`

Comment (miaulalala) [reply lib/Db/TagMapper.php:175]: Line 178

File: lib/Db/TagMapper.php [hunk 28]
```diff
@@ -0,0 +1,250 @@
...
+					$qb->expr()->eq('id', $qb->createNamedParameter($id, IQueryBuilder::PARAM_INT)),
+					$qb->expr()->eq('user_id', $qb->createNamedParameter($userId))
+				);
+		return $this->findEntity($qb);
+	}
+
+	/**
+	 * @param integer $id
+	 * @param string $userId
+	 * @return array
```
Comment (ChristophWurst) [inline lib/Db/TagMapper.php:94]:
```suggestion
	 * @return Tag[]
```

File: lib/Db/TagMapper.php [hunk 29]
```diff
@@ -0,0 +1,250 @@
...
+			->where(
+					$qb->expr()->eq('user_id', $qb->createNamedParameter($userId))
+				);
+		return $this->findEntities($qb);
+	}
+
+
+	/**
+	 * @param integer $id
+	 * @param string $userId
```
Comment (ChristophWurst) [inline lib/Db/TagMapper.php:111]: does not exist

File: lib/Db/TagMapper.php [hunk 30]
```diff
@@ -0,0 +1,250 @@
...
+			->from($this->getTableName())
+			->where(
+					$qb->expr()->eq('user_id', $qb->createNamedParameter($userId))
+				);
+		return $this->findEntities($qb);
+	}
+
+
+	/**
+	 * @param integer $id
```
Comment (ChristophWurst) [inline lib/Db/TagMapper.php:110]:
```suggestion
	 * @param int $id
```

File: lib/Db/TagMapper.php [hunk 31]
```diff
@@ -0,0 +1,250 @@
...
+
+	/**
+	 * Tag a message in the DB
+	 *
+	 * To tag (flag) a message on IMAP, @see OCA\Mail\Service\MailManager::tagMessage
+	 *
+	 * @param Tag $tag
+	 * @param string $messageId
+	 * @param string $userId
+	 * @return void
```
Comment (ChristophWurst) [inline lib/Db/TagMapper.php:132]: use the php `: void` notation instead

File: lib/Db/TagMapper.php [hunk 32]
```diff
@@ -0,0 +1,250 @@
...
+		$qb->select('*')
+			->from($this->getTableName())
+			->where($qb->expr()->eq('id', $qb->createNamedParameter($id, IQueryBuilder::PARAM_INT)));
+		return $this->findEntity($qb);
+	}
+
+	/**
+	 * Tag a message in the DB
+	 *
+	 * To tag (flag) a message on IMAP, @see OCA\Mail\Service\MailManager::tagMessage
```
Comment (ChristophWurst) [inline lib/Db/TagMapper.php:127]:
```suggestion
	 * To tag (flag) a message on IMAP, @see \OCA\Mail\Service\MailManager::tagMessage
```

then IDEs like phpstorm will let us jump directly to the method

File: lib/Db/TagMapper.php [hunk 33]
```diff
@@ -0,0 +1,250 @@
...
+		$qb->setValue('tag_id', $qb->createNamedParameter($tag->getId(), IQueryBuilder::PARAM_INT));
+		$qb->execute();
+	}
+
+	/**
+	 * Remove a tag from a DB message
+	 *
+	 * @param Tag $tag
+	 * @param string $messageId
+	 * @return void
```
Comment (ChristophWurst) [inline lib/Db/TagMapper.php:155]: `: void` on the method

File: lib/Db/TagMapper.php [hunk 34]
```diff
@@ -0,0 +1,250 @@
...
+		$qb = $this->db->getQueryBuilder();
+		$qb->delete('mail_message_tags')
+			->where($qb->expr()->eq('imap_message_id', $qb->createNamedParameter($messageId)))
+			->where($qb->expr()->eq('tag_id', $qb->createNamedParameter($tag->getId())));
+		$qb->execute();
+	}
+
+	/**
+	 * @param array $messages
+	 * @return array
```
Comment (ChristophWurst) [inline lib/Db/TagMapper.php:167]:
```suggestion
	 * @return Tag[][]
```

File: lib/Db/TagMapper.php [hunk 35]
```diff
@@ -0,0 +1,250 @@
...
+	public function untagMessage(Tag $tag, string $messageId) {
+		$qb = $this->db->getQueryBuilder();
+		$qb->delete('mail_message_tags')
+			->where($qb->expr()->eq('imap_message_id', $qb->createNamedParameter($messageId)))
+			->where($qb->expr()->eq('tag_id', $qb->createNamedParameter($tag->getId())));
+		$qb->execute();
+	}
+
+	/**
+	 * @param array $messages
```
Comment (ChristophWurst) [inline lib/Db/TagMapper.php:166]:
```suggestion
	 * @param Message[] $messages
```

File: lib/Db/TagMapper.php [hunk 36]
```diff
@@ -0,0 +1,250 @@
...
+	 * Create some default system tags
+	 *
+	 * This is designed to be similar to Thunderbird's email tags
+	 * $label1 to $label5 with the according states and colours
+	 *
+	 * <i>The array_udiff can be removed and the insert warpped in
+	 * an exception as soon as NC20 is not supported any more</i>
+	 *
+	 * @param MailAccount $account
+	 * @return void
```
Comment (ChristophWurst) [inline lib/Db/TagMapper.php:203]: move to the method signature

File: lib/Migration/Version1100Date20210304143008.php [hunk 37]
```diff
@@ -0,0 +1,121 @@
...
+		return $schema;
+	}
+
+	/**
+	 * @param IOutput $output
+	 * @param Closure $schemaClosure The `\Closure` returns a `ISchemaWrapper`
+	 * @param array $options
+	 */
+	public function postSchemaChange(IOutput $output, Closure $schemaClosure, array $options): void {
+		// repair step!
```
Comment (ChristophWurst) [inline lib/Migration/Version1100Date20210304143008.php:115]: I think you can drop this, we have a ticket for the todo

File: lib/Model/IMAPMessage.php [hunk 38]
```diff
@@ -707,10 +716,58 @@ public function toDbMessage(int $mailboxId): \OCA\Mail\Db\Message {
...
+	/**
+	 * Build tag entities from keywords sent by IMAP
+	 *
+	 * Will use IMAP keyword '$xxx' to create a value for
+	 * display_name like 'xxx'
+	 *
+	 * @link https://github.com/nextcloud/mail/issues/25
+	 *
+	 * @param array $tags
+	 * @return array
```
Comment (ChristophWurst) [inline lib/Model/IMAPMessage.php:756]:
```suggestion
	 * @return Tag[]
```

File: lib/Model/IMAPMessage.php [hunk 39]
```diff
@@ -707,10 +716,58 @@ public function toDbMessage(int $mailboxId): \OCA\Mail\Db\Message {
...
+
+	/**
+	 * Build tag entities from keywords sent by IMAP
+	 *
+	 * Will use IMAP keyword '$xxx' to create a value for
+	 * display_name like 'xxx'
+	 *
+	 * @link https://github.com/nextcloud/mail/issues/25
+	 *
+	 * @param array $tags
```
Comment (ChristophWurst) [inline lib/Model/IMAPMessage.php:755]:
```suggestion
	 * @param string[] $tags
```

**Review by ChristophWurst — CHANGES_REQUESTED** [review]
Proper types and their documentation help developers, IDEs and static analysis :)

File: lib/Db/TagMapper.php (hunk 36 above)
Comment (miaulalala) [reply lib/Db/TagMapper.php:203]: ?

File: lib/Db/TagMapper.php (hunk 27 above)
Comment (miaulalala) [reply lib/Db/TagMapper.php:175]: Nah it's fine, it was just a superfluous assignment.

**Review by ChristophWurst — APPROVED** [review]
Tested and works :shipit: 

Now you may do the final rebase & squash
